From 998bfb87b535dcd3ad739accd720642fecc7f0cf Mon Sep 17 00:00:00 2001
From: Stefan Chulski <stefanc@marvell.com>
Date: Tue, 28 Jun 2016 17:36:47 +0300
Subject: [PATCH 0541/2241] net: mvpp2x: fix issue that extra refill would be
 done to BM pool

- extra buffer would be refilled to BM pool when skb build failed

Change-Id: Ibe50066596afcad752e2fc1060038ec14ea8660f
Signed-off-by: Stefan Chulski <stefanc@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/30797
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Reviewed-by: Yuval Caduri <cyuval@marvell.com>
---
 drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c | 14 ++++++--------
 1 file changed, 6 insertions(+), 8 deletions(-)

diff --git a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
index 59a9ec1..efd1fa4 100644
--- a/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
+++ b/drivers/net/ethernet/marvell/mvpp2x/mv_pp2x_main.c
@@ -2243,21 +2243,19 @@ err_drop_frame:
 				data);
 			continue;
 		}
-		atomic_inc(&bm_pool->in_use);
-		refill_array[bm_pool->log_id]++;
 
 		skb = build_skb(data, bm_pool->frag_size > PAGE_SIZE ? 0 : bm_pool->frag_size);
-		/* After refill old buffer has to be unmapped regardless if
-		 * the skb is successfully built or not.
-		 */
-		dma_unmap_single(dev->dev.parent, buf_phys_addr,
-				 MVPP2_RX_BUF_SIZE(bm_pool->pkt_size), DMA_FROM_DEVICE);
-
 		if (!skb) {
 			pr_err("skb build failed\n");
 			goto err_drop_frame;
 		}
 
+		dma_unmap_single(dev->dev.parent, buf_phys_addr,
+				 MVPP2_RX_BUF_SIZE(bm_pool->pkt_size), DMA_FROM_DEVICE);
+
+		atomic_inc(&bm_pool->in_use);
+		refill_array[bm_pool->log_id]++;
+
 #ifdef MVPP2_VERBOSE
 		mv_pp2x_skb_dump(skb, rx_desc->data_size, 4);
 #endif
-- 
2.7.4

